packages/swingset-liveslots/src/boyd-gc.js

```diff
@@ -27,138 +28,393 @@ import { parseVatSlot } from './parseVatSlots.js';
+  `exportedRemotables` set. And when it is stored in vdata, it appears
+  as a key of the `remotableRefCounts` map. That keeps the Remotable
+  itself alive until the other reachability pathways have gone
+  away. We don't do this for Representatives because it would violate
+  our "don't use RAM for inactive virtual objects" rule.
+
+  When an imported Presence becomes UNREACHABLE, it might still be
+  RECOGNIZABLE, by virtue of being the key of one or more weak
+  collections. If not, it might transit from REACHABLE directly to
+  NONE. The code that reacts to the UNREACHABLE transition must check
+  for recognizers, and do a retireImports right away if there are
+  none. Otherwise, recognizer will remain until either the kernel
+  retires the object (dispatch.retireImports), or the weak collection
+  is deleted, in which case possiblyRetiredSet will be updated with
+  the vref that might no longer be recognized. There will be a race
+  between us ceasing to recognize the vref (which should trigger a
+  syscall.retireImports), and the kernel telling us the object has
+  been deleted (via dispatch.retireImports). Either one must inhibit
+  the other.
+
+  possiblyRetiredSet only cares about Presence-style vrefs, because
+  they represent imports, whose lifetime is not under our control. The
+  collection-deletion code will add Remotable- and Representative-
+  style vrefs in possiblyRetiredSet, but we can remove and ignore
+  them.
+
+  We use slotToVal.has(vref) everywhere for our "is it still
+  reachable" check, which returns true for the REACHABLE / UNREACHABLE
+  / COLLECTED states, and false for the FINALIZED state. In contrast,
+  getValForSlot(vref) returns false for both COLLECTED and
+  FINALIZED. We want COLLECTED to qualify as "still reachable" because
+  it means there's still a finalizer callback queued, which will be
+  run eventually, and we need that callback to not trigger a duplicate
+  drop. We use slotToVal.has() in the possiblyRetiredSet loop (to
+  inhibit retirement of imported vrefs whose Presences are in the
+  COLLECTED state, and which just lost a recognizer), because
+  getValForSlot would allow such COLLECTED vrefs to be retired, even
+  before the finalizer had fired and could do a dropImports.
 
+  Deleting a virtual object will delete its `state`, decrementing the
+  refcounts of any objects therein, which might shake loose more
+  data. So we keep looping until we've stopped adding things to
+  possiblyDeadSet. Retiring a Presence-style vref, on the other hand,
+  cannot shake anthing loose. So we do not include the
+  possiblyRetiredSet processing inside that loop.
+
+  When the kernel sends us a dispatch.bringOutYourDead (or "BOYD" for
+  short), this scanForDeadObjects() will be called. This is the only
+  safe time to do anything that is sensitive to engine GC behavior,
+  like looking inside WeakRefs, or reading sets that are modified
+  during FinalizationRegistry callbacks. So this is the only time we
```

Comment:
Hm, good point, so that would be when we unmarshal some arguments, and we get an `o-4`, and we need a Presence, and `provideValForSlot()` must create the Presence if either `slotToVal.get('o-4')` comes up empty (FINALIZED), or if it yields a WeakRef but the WeakRef is empty (COLLECTED). If `slotToVal.get(vref)?.deref()` comes up with a Presence (REACHABLE or UNREACHABLE) then it doesn't need to create one.

We went through a couple of rounds of development to make sure the syscalls performed do not vary between those two cases. For Presences it's easy, we don't need any syscalls to build a Presence. For Representatives, we don't need the `syscall.vatstoreGet` until we need `state`, and we don't need that until someone invokes a method, and *that* is unaffected by GC state.

So yeah, we must peek into the WeakRef, but we don't do any syscalls in response to it, and that's as far as we think we can manage. Metering might be affected (modulo us disabling metering during unmarshalling, which I think we kinda half do right now), but not syscalls.

Let me modify this clause to point out that "safe" means "we're allowed to let the syscall trace be influenced by GC state now".

Reply:
new text:


  When the kernel sends us a dispatch.bringOutYourDead (or "BOYD" for
  short), this scanForDeadObjects() will be called. This is the only
  appropriate time for the syscall behavior to depend upon engine GC
  behavior: during all other deliveries, we want the syscalls to be a
  deterministic function of delivery contents, userspace behavior, and
  vatstore data.

  During BOYD, we still try to minimize the variation of behavior as
  much as possible. The first step is to ask the engine to perform a
  full GC sweep, to collect any remaining UNREACHABLE objects, and
  allow the finalizer callbacks to run before looking at the
  results. We also sort the vrefs before processing them, to remove
  sensitivity to the exact order of finalizer invocation.

  That makes BOYD a safe time to look inside WeakRefs and make
  syscalls based on the contents, or to read the sets that are
  modified during FinalizationRegistry callbacks and make syscalls to
  query their state further. This this is the only time we examine and
  clear possiblyDeadSet and possiblyRetiredSet, or probe
  slotToVal. Outside of BOYD, in convertSlotToVal, we must probe the
  WeakRefs to see whether we must build a new Presence or
  Representative, or not, but we have carefully designed that code to
  avoid making syscalls during the unmarshalling process, so the only
  consequence of GC differences should be differences in metering and
  memory allocation patterns.